openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -71,42 +76,42 @@ def assert_user_in_session(self):
         """
         assert SafeSessionMiddleware.get_user_id_from_session(self.request) == self.user.id
 
-    @patch("openedx.core.djangoapps.safe_sessions.middleware.LOG_REQUEST_USER_CHANGES", False)
     @patch("openedx.core.djangoapps.safe_sessions.middleware.log_request_user_changes")
-    def test_success(self, mock_log_request_user_changes):
-        self.client.login(username=self.user.username, password='test')
-        session_id = self.client.session.session_key
-        safe_cookie_data = SafeCookieData.create(session_id, self.user.id)
-
-        # pre-verify steps 3, 4, 5
-        assert getattr(self.request, 'session', None) is None
-        assert getattr(self.request, 'safe_cookie_verified_user_id', None) is None
-
-        # verify step 1: safe cookie data is parsed
-        self.assert_response(safe_cookie_data)
-        self.assert_user_in_session()
-
-        # verify step 2: cookie value is replaced with parsed session_id
-        assert self.request.COOKIES[settings.SESSION_COOKIE_NAME] == session_id
-
-        # verify step 3: session set in request
-        assert self.request.session is not None
-
-        # verify steps 4, 5: user_id stored for later verification
-        assert self.request.safe_cookie_verified_user_id == self.user.id
-
-        # verify extra request_user_logging not called.
-        assert not mock_log_request_user_changes.called
-
-    @patch("openedx.core.djangoapps.safe_sessions.middleware.LOG_REQUEST_USER_CHANGES", True)
-    @patch("openedx.core.djangoapps.safe_sessions.middleware.log_request_user_changes")
-    def test_log_request_user_on(self, mock_log_request_user_changes):
-        self.client.login(username=self.user.username, password='test')
-        session_id = self.client.session.session_key
-        safe_cookie_data = SafeCookieData.create(session_id, self.user.id)
-
-        self.assert_response(safe_cookie_data)
-        assert mock_log_request_user_changes.called
+    @ddt.unpack
+    @ddt.data(
+        {'toggle': True, 'cache_flag': True, 'should_log': True},
+        {'toggle': True, 'cache_flag': False, 'should_log': False},
+        {'toggle': False, 'cache_flag': True, 'should_log': False},
+        {'toggle': False, 'cache_flag': False, 'should_log': False},
+    )
+    def test_success(self, mock_log_request_user_changes, toggle, cache_flag, should_log):
+        with patch("openedx.core.djangoapps.safe_sessions.middleware.LOG_REQUEST_USER_CHANGES", toggle):
+            self.client.login(username=self.user.username, password='test')
+            cache.set(USER_MISMATCH_CACHE_KEY, cache_flag)
```

Comment:
It feels like we should assert `USER_MISMATCH_CACHE_KEY` at the end as well. At this time, does a normal login trigger a mismatch? Would the cache always be True? Maybe this will change after Tim's work, but would capture how it should be working now.